common/persistence/client/factory.go

```diff
@@ -246,13 +249,81 @@ func (f *factoryImpl) NewExecutionManager(shardID int) (p.ExecutionManager, erro
 }
 
 // NewVisibilityManager returns a new visibility manager
-func (f *factoryImpl) NewVisibilityManager() (p.VisibilityManager, error) {
-	visConfig := f.config.VisibilityConfig
+func (f *factoryImpl) NewVisibilityManager(
+	params *service.BootstrapParams,
+	resourceConfig *dynamicconfig.ResourceConfig,
+) (p.VisibilityManager, error) {
+	if resourceConfig.EnableReadVisibilityFromES == nil && resourceConfig.AdvancedVisibilityWritingMode == nil {
+		// No need to create visibility manager as no read/write needed
+		return nil, nil
+	}
+	var visibilityFromDB, visibilityFromES p.VisibilityManager
+	var err error
+	if params.PersistenceConfig.VisibilityStore != "" {
+		visibilityFromDB, err = f.newDBVisibilityManager(resourceConfig)
+		if err != nil {
+			return nil, err
+		}
+	}
+	if params.PersistenceConfig.AdvancedVisibilityStore != "" {
+		visibilityIndexName := params.ESConfig.Indices[common.VisibilityAppName]
+		visibilityProducer, err := params.MessagingClient.NewProducer(common.VisibilityAppName)
+		if err != nil {
+			f.logger.Fatal("Creating visibility producer failed", tag.Error(err))
```

Comment:
why not return error but log fatal?

Reply:
It was copied it from https://github.com/uber/cadence/blob/723ecf572b2391ad3acc93c455445d9968170eef/service/history/service.go#L83 I am fine with either way. Returning error should result in the same behavior.

Reply:
Lots of our code is using fatal.
Some thoughts: this fatal would result in a panic and even the caller ignores the error, the startup would still fail. Also, this can report a "Fatal" log type instead of "error". which is more obvious for us to catch(since we may have many other errors). And this one single line is good for :
```
logger.Error(...)
return err
```

Do you feel okay to keep Fatal?

Reply:
I'm okay with it.